### PR TITLE
BUG: Fix numba DUFuncs added loops getting picked up

### LINKED ISSUE #20735: BUG:  ufunc 'x' did not contain a loop with signature matching types
### Describe the issue:

The following code started throwing this error after upgrading from numpy 1.21.2 to 1.22.0.
Any help would be much appreciated.

### Reproduce the code example:

```python
import numba
import numpy

@numba.vectorize(nopython=True, cache=True)
def v_floor(a):
    return numpy.floor(a)

v_floor(1.9)
```

### Error message:

```shell
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
<ipython-input-4-200f7bdc40b0> in <module>
----> 1 v_floor(1.9)

UFuncTypeError: ufunc 'v_floor' did not contain a loop with signature matching types <class 'numpy.dtype[float64]'> -> None
```

### NumPy/Python version information:

python                    3.9.9
numba                     0.53.1
numpy                     1.22.0

### PR DESCRIPTION
**Pull request by seberg**
It was always my intention to do this: If no loop is found and we
go into the legacy ufunc path and legacy resolving works, we need
to double check that the ufunc was not mutated.

Normal operation never mutates ufuncs (its really not meant to be)
but numbas DUFuncs need to do it (they compile loops dynamically).
The future is much brighter for them in this regard, but right now
they have to keep working.

Closes gh-20735

### CODE REVIEW

**Comment by seberg** [comment]
Note: This needs to be tested by numba, but I currently don't have a working numba environment.  (The test doesn't have to be thorough, if any DUFunc works, we should be good.)

File: numpy/core/src/umath/dispatching.c [hunk 1]
```diff
@@ -746,6 +746,37 @@ promote_and_get_info_and_ufuncimpl(PyUFuncObject *ufunc,
...
+        npy_bool loop_exists = NPY_FALSE;
+        for (int i = 0; i < ufunc->ntypes; ++i) {
+            loop_exists = NPY_TRUE;  /* assume it exists, break if not */
+            for (int j = 0; j < ufunc->nargs; ++j) {
+                if (types[j] != new_op_dtypes[j]->type_num) {
+                    loop_exists = NPY_FALSE;
+                    break;
+                }
+            }
+            types += ufunc->nargs;
```
Comment (charris) [inline numpy/core/src/umath/dispatching.c:771]: Why doesn't the loop exit if a compatible loop has been found? I'm missing something.

Comment (seberg) [reply numpy/core/src/umath/dispatching.c:774]: nah, you are not missing anything, I was just being blind :).   I stole the code, and the code was smarter than me, used the "long" solution for now, the original code continued checking `j == ufunc->nargs`...

Anyway, tested the code manually, now.  So I think it is OK and should be good.

**Comment by charris** [comment]
Thanks Sebastian.

**Pull request merged**